R/add.R

```diff
@@ -301,20 +257,45 @@ m_add_line <- function(
 #'       m_style_sphere(scale = 0.3)
 #'     )
 #'   ) %>%
-#'   m_add_lines(line_specs = line_specs)
+#'   m_add_lines(
+#'     starts = m_sel(resi = 1, chain = "A"),
+#'     ends = list(
+#'       m_sel(resi = 10, chain = "A"),
+#'       m_sel(resi = 10, chain = "B")
+#'     ),
+#'     dashed = TRUE
+#'   )
 m_add_lines <- function(
                         id,
-                        line_specs) {
-  if (missing(line_specs) || is.null(line_specs)) {
-    stop("`line_specs` should be passed in.")
+                        starts,
+                        ends,
+                        dashed = FALSE,
+                        pairwise = FALSE,
+                        spec = m_shape_spec()) {
```

Comment:
If we expand all the args and remove the `m_add_line()`, users are not be able to create multiple lines with different style (dashed or spec). I think something like `m_line_spec()` function which  returns a list contians `start`, `end`, `dashed`, `spec` would be more general?

Reply:
I am unsure if adding a whole new function for `m_line_spec()` is the right approach. 

The complexity increases a lot when trying to create multiple lines, with lots of different colours. 

I think the best approach might just be passing in the `dashed`, `color` etc values as vectors into the `m_add_lines()` to get different values for the lines. THe vector being passed in will just have to be the same length (or length == 1) for the number of lines being created.

Example: 
```r
m_add_lines(
  starts = list(
    m_sel(resi = 1), 
    m_sel(resi = 2)
    ), 
  ends = list(
    m_sel(resi = 3), 
    m_sel(resi = 4)
    ), 
  dashed = c(TRUE, FALSE), 
  color = c("blue", "red")
)
```
This would creat a blue dashed line between 1 & 3, and a red solid line between 2 & 4.

Reply:
I have just done a test implementation now: 
```r
r3dmol() %>% 
  m_add_model(m_fetch_pdb("1bna")) %>% 
  m_set_style(m_style_cartoon()) %>% 
  m_add_lines(
    starts = m_multiresi_sel(resi = 1:3, chain = "A"), 
    ends = m_multiresi_sel(resi = 22:20, chain = "B"), 
    dashed = TRUE, 
    color = c("blue", "red", "green")
  ) %>% 
  m_zoom_to()
```
<p align="center">
<img src="https://user-images.githubusercontent.com/36021261/108951024-59a6d600-76bb-11eb-9954-edf1be62d2ec.png" width="400"></P>

As long as the aesthetic values are either `length() == 1` in which case they are recycled for every line, or `length() == #number_of_lines` then it works out.

Reply:
Yes, this approach is possible. My concern is that if lots of atoms are selected at the same time, there is no strong restrictive association between the options in one selected atom, so it might be easy to make mistakes. If pass in a data.frame that contains columns with options instead of a list, it's easier to find out if there are any mistakes. 

For example something like:
```r
df <- data.frame(
  starts = 1:2,
  ends = 3:4,
  dashed = c(TRUE, FALSE),
  color = c("blue", "red")
)

# options set of the first atom
df[1, ]
```

Reply:
The problem I see with this is that the complex will have to be quite complex. If you want a `data.frame` to contain all of your options, you will need a `starts.chain`, `ends.chain`, `starts.resi` `etc.` for every possible selection option. I feel like this is a more complex and potentially confusing way of going about it rather than just passing in lists of selections.

Reply:
Yes, using list is a good idea. I think it would be better to use the long structure instead of wide structure.
For example, if using the wide structure:
```r
# Method 1
m_add_lines(
    starts = 1:10, 
    ends = 11:20, 
    dashed = c(T, F, T, F, F, F, T, T, T, F)
    color = c("blue", "red", "green", "red", "green",  "red", "green",  "red", "green", "blue")
  ) %>% 
```
It would be very difficult to check if the options are what the user expected.
It would be more easy to understand if the options of each atom is in an element.
```r
# Method 2
list(
  list(starts = 1, ends = 11, dashed = T, color = "blue"),
  list(starts = 2, ends = 12, dashed = F, color = "red"),
  list(starts = 3, ends = 13, dashed = T, color = "green")
)
```
In method 1, although it is possible to restrict by checking the length of each options, it is not possible to check whether each setting is in the right index that the user expected.

Reply:
You are correct that it becomes difficult to check whether the options match up with the index that the user wants, but otherwise doing it as a list as `# Method 2`
 >```r
># Method 2
>list(
>  list(starts = 1, ends = 11, dashed = T, color = "blue"),
>  list(starts = 2, ends = 12, dashed = F, color = "red"),
>  list(starts = 3, ends = 13, dashed = T, color = "green")
>)
>```

Is essentially the same as: 
```r
m_add_lines(starts = 1, ends = 11, dashed = T, color = "blue") %>%
m_add_lines(starts = 2, ends = 12, dashed = F, color = "red") %>%
m_add_lines(starts = 3, ends = 13, dashed = T, color = "green")
```

So users can always call multiple `m_add_lines()` one after the other, but it would be best to allow higher-throughput entry into the `m_add_lines()` function itself, to allow for easier bulk-line creation.

Reply:
I don't think that manually using `m_add_lines()` will be used very often. 

I see it being used to import the results from a different analysis. In that case, the results would already be pre-formatted and matched up, and it's just a matter of getting it into the `m_add_lines()` function as easily as possible. 

It would likely just allow for better interfacing with the outputs of packages like `{bio3d}` and others.

Reply:
Thanks, I think these two reasons are convincing
1. I see it being used to import the results from a different analysis...It would likely just allow for better interfacing with the outputs of packages like {bio3d} and others.
2. Is essentially the same as...

Although I pointed out earlier that changing `m_add_line` to an internal function `.m_add_line` and add `m_add_lines`, the interface between `m_add_lines` and `m_add_line` are the same. To prevent breaking changes, it would be better to rename `m_add_lines` back to `m_add_line`. The difference between the previous version is simply that it can accept more atoms selection at once.

Reply:
Sounds good. I"ll go through and make the changes.

Reply:
Thanks for all of the really great feedback and discussion. I really appreciate it!